wgpu-hal/Cargo.toml

```
@@ -144,6 +147,8 @@ khronos-egl = { version = "6", features = ["static", "no-pkg-config"] }
 libloading = { version = ">=0.7, <0.9", optional = true }
 
 [target.'cfg(windows)'.dependencies]
+# backend: Dx12 and Gles
+windows = { version = "0.58", default-features = false, optional = true }
```

since there's more `windows` crate uses coming, let's put that into the workspace `Cargo.toml` (make sure it's also set with default-features false in there, otherwise it opts in default features for everything with now way to opt out)

```suggestion
windows = { workspace = true, optional = true }
```

---

As before, I'm still confused by this unused yet _partially_ maintained workspace dependencies table :upside_down_face: 

Why are some but not all dependencies inherited from the workspace?

---

Also, do folks want to start with Windows 0.57 or Windows 0.58? The latter has MSRV 1.70 but it seems `wgpu` is already on `1.76`.

Note that `default-features` only turns off the `std` feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.

---

>  Why are some but not all dependencies inherited from the workspace?

wgpu is older than the ability to do workspace inheritance of crates and nobody every moved stuff over consistently 🤷 

> Also, do folks want to start with Windows 0.57 or Windows 0.58? The latter has MSRV 1.70 but it seems wgpu is already on 1.76.

Any reason in favor of Windows 0.57? As you pointed out from MSRV pov that's not an issue, so we should rather stay up to date? If there's a reason to stick to the older we also have to make this known to dependabot, otherwise an update will creep in latest next week ;-) 

> Note that default-features only turns off the std feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.

Good to know, maybe worth a comment on it in case you can confirm? Generally good practice for a library like wgpu not to drag in default features when they're not needed even if it's something like improved error handling, after all a user of wgpu can still do that just fine. But the other way round a user of wgpu wouldn't be able to opt-out of those default features at all

---

> wgpu is older than the ability to do workspace inheritance of crates and nobody every moved stuff over consistently 🤷

It was previously told to me that workspace dependency inheritance was reverted, but the workspace table was kept in place so that it could ever get reenabled.

> Any reason in favor of Windows 0.57? As you pointed out from MSRV pov that's not an issue, so we should rather stay up to date? If there's a reason to stick to the older we also have to make this known to dependabot, otherwise an update will creep in latest next week ;-)

None really; except that I started the conversion on `0.57`. A lot of _my_ high-level API improvements in `Dxgi` via metadata annotations finally landed in `windows 0.58`, too. All dependencies like `gpu-allocator` have released with `windows 0.58` in their range as well.

---

I'll double-check the purpose of the `std` feature before marking this thread as resolved.

---

For context on the workspace table, we introduced it, but there was a bug in cargo-vendor that caused Firefox to not be able to injest it, so we reverted core, hal, and types (the only crates ff uses) back to standard dependencies. The bug has been long fixed, I just don't think anyone has put in the work to verify it's fine and move back to a workspace dep.

---

I've opened #6020 to start cleaning up some of the workspace dependencies (if that's helpful).

---

Thanks @waywardmonkeys, looking forward to see that PR fix at least the dependencies used here, so that this PR can simply be based on top. Looks like you have a lot of dependencies to go, though.

---

> > Note that default-features only turns off the std feature: I'm not exactly sure what that implies but we may get worse error formatting - at least it was like that years ago.
> 
> Good to know, maybe worth a comment on it in case you can confirm? Generally good practice for a library like wgpu not to drag in default features when they're not needed even if it's something like improved error handling, after all a user of wgpu can still do that just fine. But the other way round a user of wgpu wouldn't be able to opt-out of those default features at all

Upon closer inspection, the feature is broken.

- `windows-core`: Only uses it to turn on `Event`, which is a wrapper for delegates. The feature is **not forwarded to windows-result` nor `windows-strings`** (which is a bug I'd say);
- `windows`: only forwards it to `windows-core`.

The transitive crates:

- `windows-result`: implements `std::error::Error` for `Error`, and conversions to/from `std::io::Error` (which is very similar on Windows);
- `windows-strings`: various conversion for `HSTRING` to and from `OsStr(ing)` and `Path`.

In the end it's nothing that seems to be hiding nicer internals such as error formatting from us. Everything that it exposes is explicit public API (and if we end up relying on it, turning off `std` will cause compiler errors).